third_party/gfxreconstruct/framework/decode/dive_block_data.cpp

```diff
@@ -0,0 +1,160 @@
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+bool DiveBlockData::AddOriginalBlock(uint32_t index, uint64_t offset)
+{
+    if (original_blocks_map_locked_)
+    {
+        GFXRECON_LOG_ERROR("Cannot add more original blocks to a locked DiveBlockData object");
+        return false;
+    }
+
+    if (index != original_blocks_map_.size())
+    {
+        GFXRECON_LOG_ERROR(
+            "Unexpected block id mismatch with index: %d, expected index: %d", index, original_blocks_map_.size());
+        return false;
+    }
+
+    original_blocks_map_.push_back(BlockBytesLocation{ offset, 0 });
+
+    return true;
+}
+
+bool DiveBlockData::LockOriginalBlocksMap()
+{
+    if (original_blocks_map_locked_)
+    {
+        GFXRECON_LOG_INFO("Original block map already locked");
+        return true;
+    }
+
+    if (original_blocks_map_.size() == 0)
+    {
+        GFXRECON_LOG_ERROR("Original block map is empty");
+        return false;
+    }
+
+    // Calculating block sizes
+    original_header_size_bytes_ = original_blocks_map_[0].offset;
+    for (int i = 0; i < original_blocks_map_.size() - 1; i++)
+    {
+        uint64_t size                = original_blocks_map_.at(i + 1).offset - original_blocks_map_.at(i).offset;
+        original_blocks_map_[i].size = size;
+    }
+
+    // Removing the last "block" which is the end of the file
+    original_blocks_map_.pop_back();
+
+    original_blocks_map_locked_ = true;
+    return true;
+}
+
+bool DiveBlockData::GetNewBlocksOrder()
+{
+    new_blocks_order_.clear();
+
+    // TODO(chenangela): Insert modified blocks in the correct order
+    for (uint32_t primary_index = 0; primary_index < original_blocks_map_.size(); primary_index++)
+    {
+        NewBlockMetadata block;
+        block.is_original = true;
+        block.id          = primary_index;
+        new_blocks_order_.push_back(block);
+    }
+
+    return true;
+}
+
+bool DiveBlockData::WriteGFXRFile(const std::string& original_file_path, const std::string& new_file_path)
+{
+    if (!original_blocks_map_locked_)
+    {
+        GFXRECON_LOG_ERROR("DiveBlockData original map must be locked before writing new file");
+        return false;
+    }
+
+    bool res = GetNewBlocksOrder();
+    if (!res)
+    {
+        GFXRECON_LOG_ERROR("Cannot determine new blocks order");
+        return false;
+    }
+
+    FILE* original_fd;
+    int   result = util::platform::FileOpen(&original_fd, original_file_path.c_str(), "rb");
+    if (result || original_fd == nullptr)
+    {
+        GFXRECON_LOG_ERROR("Failed to open file %s", original_file_path.c_str());
+        return false;
+    }
+
+    FILE* new_fd;
+    result = util::platform::FileOpen(&new_fd, new_file_path.c_str(), "wb");
+    if (result || new_fd == nullptr)
+    {
+        GFXRECON_LOG_ERROR("Failed to open file %s", new_file_path.c_str());
+        return false;
+    }
+
+    // Copy the original header
+    util::platform::FileRead(block_buffer_, original_header_size_bytes_, original_fd);
```

Comment:
is there a possibility that `BUFFER_SIZE` is smaller than `original_header_size_bytes_`? maybe better with an assert?

Reply:
sorry i don't see the new change on this. is it possible that the latest version is not committed correct?

Reply:
thanks